bam_checksum.c

```diff
@@ -0,0 +1,1311 @@
+                    header[i] = H_POS, o->check_pos = 1;
+                else if (strcmp(token, "+cigar") == 0)
+                    header[i] = H_CIGAR, o->check_cigar = 1;
+                else if (strcmp(token, "+mate") == 0)
+                    header[i] = H_MATE, o->check_mate = 1;
+                else if (strcmp(token, "combined") == 0)
+                    header[i] = H_COMBINED;
+                else  {
+                    fprintf(stderr, "Unrecognised header token '%s'\n", token);
+                    goto err;
+                }
+
+                i++;
+                ptr += idx;
+            }
+            nheader = i;
+
+            continue;
+        }
+
+        if (strncmp(line.s, "# Aux", 5) == 0) {
+            int idx;
+            char c;
+            if (sscanf(line.s, "# Aux tags: %c%n", &c, &idx) == 1)
+                if (!o->tag_str)
+                    o->tag_str = strdup(line.s + idx-1);
+
+            continue;
+        }
+
+        if (strncmp(line.s, "# BAM", 5) == 0) {
+            int idx;
+            char c;
+            if (sscanf(line.s, "# BAM flags: %c%n", &c, &idx) == 1)
+                o->flag_mask = bam_str2flag(line.s + idx-1);
+
+            continue;
+        }
+
+        if (!line.l || *line.s == '#')
+            continue;
+
+
+        // Header done.  Now parse the data lines
+        if (strncmp(line.s, "all ", 4) == 0 ||
+            strncmp(line.s, "all\t", 4) == 0)
+            continue;
+
+        char col[11][128], *ptr = line.s;
+        int nf;
+        for (nf = 0; nf < 11; nf++) {
+            int idx;
+            int n = sscanf(ptr, "%127s%n", col[nf], &idx);
+            if (n <= 0)
+                break;
+            if (strlen(col[nf]) == 127) {
+                fprintf(stderr, "Field too long\n");
+                goto err;
+            }
+            ptr += n == 1 ? idx : 0;
+        }
+
+        // Sanity check that header and rows match
+        if (nf < 8 || nf != nheader) {
+            fprintf(stderr, "Incorrect number of columns in line: %s\n",
+                    line.s);
+            goto err;
+        }
+
+        // Marry up column header with row entries and set struct.
+        // (We could update the struct to be numbered instead of
+        // named in variables to make this easier.)
+        int qc = 0;
+        uint64_t count = 0;
+        for (int i = 0; i < nf; i++) {
+            switch (header[i]) {
+            case H_QC:
+                if (strcmp(col[i], "all") == 0)
+                    qc = 0;
+                else if (strcmp(col[i], "pass") == 0)
+                    qc = 1;
+                else if (strcmp(col[i], "fail") == 0)
+                    qc = 2;
+                else
+                    goto err;
+                break;
+
+            case H_COUNT:
+                count = strtol(col[i], NULL, 10);
+                break;
+
+            case H_SEQ:
+                c.seq = strtol(col[i], NULL, 16);
```

Comment:
Should this be `strtoul()` to prevent values >= 0x8000000 being converted to `LONG_MAX` on platforms with 32-bit longs?  Or can that never happen due to the prime used in the summing algorithm?